File: VB/Customfigs_VB/VectorDrawCustomFigs_VB/VectorDrawPolygon.vb
Imports System
Imports System.Runtime.InteropServices
Imports System.ComponentModel
Imports VectorDraw.Serialize
Imports VectorDraw.Professional
Imports VectorDraw.Professional.vdPrimaries
Imports VectorDraw.Professional.vdFigures
Imports VectorDraw.Professional.vdCollections
Imports VectorDraw.Professional.vdObjects
Imports VectorDraw.Professional.Constants
Imports VectorDraw.GeomeTry
Imports VectorDraw.Render
Imports VectorDraw.Professional.Actions
Imports VectorDraw.Actions
'the VectorDrawCustomFigs.dll must be in the folder of vdrawcomponents
'or the supportpath must contain the path of the assemply dll

    Public Class VectorDrawPolygon
        Inherits vdShape
#Region "main variables"
        Private mTextStyle As vdTextstyle = Nothing
        Private mNumSides As Integer = 4
        Private mText As String = ""
        Private mTextHeight As Double = 1D
        Private mRadius As Double = 1D
#End Region

#Region "ctor"
        Public Sub New()
        End Sub
#End Region

#Region "not vdShape supported properties"
        <Browsable(False)> _
        Public Overrides Property Rotation() As Double
            Get
                'return base.Rotation;
                Return 0D
            End Get
            Set(ByVal Value As Double)
                'base.Rotation = value;
            End Set
        End Property
        <Browsable(False)> _
        Public Overrides Property Scales() As Vector
            Get
                'return base.Scales;
                Return New Vector(1D, 1D, 1D)
            End Get
            Set(ByVal Value As Vector)
                'base.Scales = value;
            End Set
        End Property
#End Region
#Region "vdShape override properties"
        Public Overrides Property Origin() As gPoint
            Get
                Return MyBase.Origin
            End Get
            Set(ByVal Value As gPoint)
                MyBase.Origin = Value
                Update()
            End Set
        End Property
#End Region
#Region "properties"
        'You should use the AddHistory command in your properties so any change will be writen to the history
    'The string in the AddHistory command is the name of the property exactly and should be called before the property change.
        Public Property NumSides() As Integer
            Get
                Return mNumSides
            End Get
            Set(ByVal Value As Integer)
                AddHistory("NumSides", Value)
                mNumSides = Value
            End Set
        End Property
        Public Property TextString() As String
            Get
                Return mText
            End Get
            Set(ByVal Value As String)
                AddHistory("TextString", Value)
                mText = Value
            End Set
        End Property
        Public Property Radius() As Double
            Get
                Return mRadius
            End Get
            Set(ByVal Value As Double)
                AddHistory("Radius", Value)
                mRadius = Value
            End Set
        End Property
        Public Property TextStyle() As vdTextstyle
            Get
                Return mTextStyle
            End Get
            Set(ByVal Value As vdTextstyle)
                AddHistory("TextStyle", Value)
                mTextStyle = Value
            End Set
        End Property
        Public Property TextHeight() As Double
            Get
                Return mTextHeight
            End Get
            Set(ByVal Value As Double)
                AddHistory("TextHeight", Value)
                mTextHeight = Value
            End Set
        End Property
#End Region
#Region "serialization"
    'Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
    'You can set value of custom property to any of the following type values: 
    '     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise
    Public Overrides Sub Serialize(ByVal serializer As Serializer)
        MyBase.Serialize(serializer)
        serializer.Serialize(mNumSides, "NumSides")
        serializer.Serialize(mRadius, "Radius")
        serializer.Serialize(mText, "TextString")
        serializer.Serialize(mTextStyle, "TextStyle")
        serializer.Serialize(mTextHeight, "TextHeight")
    End Sub
    Public Overrides Function DeSerialize(ByVal deserializer As DeSerializer, ByVal fieldname As String, ByVal value As Object) As Boolean
        If MyBase.DeSerialize(deserializer, fieldname, value) Then
            Return True
        ElseIf fieldname = "NumSides" Then
            mNumSides = CType(value, Integer)
        ElseIf fieldname = "Radius" Then
            mRadius = CType(value, Double)
        ElseIf fieldname = "TextString" Then
            mText = CType(value, String)
        ElseIf fieldname = "TextStyle" Then
            mTextStyle = CType(value, vdTextstyle)
        ElseIf fieldname = "TextHeight" Then
            mTextHeight = CType(value, Double)
        Else : Return False
        End If
        Return True
    End Function
#End Region
#Region "curve calulation"
        Public Overrides Sub FillShapeEntities(ByRef entities As vdEntities) 'calculate shape entities in ecs 
            Dim text As vdText = New vdText()
            entities.AddItem(text)
            text.MatchProperties(Me, Document)

            text.Style = TextStyle
            text.TextString = TextString
            text.VerJustify = VdConstVerJust.VdTextVerCen
            text.HorJustify = VdConstHorJust.VdTextHorCenter
            text.Height = TextHeight

            Dim pl As vdPolyline = New vdPolyline()
            entities.AddItem(pl)
            pl.MatchProperties(Me, Document)
            Dim verts As Vertexes = New Vertexes()
            Dim stepangle As Double = Globals.VD_TWOPI / Me.NumSides
            Dim sang As Double = 0D
            Dim cen As gPoint = New gPoint()
            Dim i As Integer
            For i = 0 To NumSides - 1 Step i + 1
                verts.Add(gPoint.Polar(cen, sang, Radius))
                sang += stepangle
            Next
            pl.VertexList = verts
            pl.Flag = VdConstPlineFlag.PlFlagCLOSE

        End Sub
#End Region
#Region "overrides"
        Protected Overrides Sub OnDocumentDefaults()
            MyBase.OnDocumentDefaults()
            mTextStyle = Document.ActiveTextStyle
            If mTextStyle Is Nothing Then
                mTextStyle = Document.TextStyles.Standard
            End If
        End Sub
        Public Overrides Sub InitializeProperties()
            mNumSides = 4
            mText = ""
            mTextHeight = 1D
            mTextStyle = Nothing
            If Not Document Is Nothing Then
                mTextStyle = Document.TextStyles.Standard
            End If
            mRadius = 1D
            MyBase.InitializeProperties()
        End Sub
        Public Overrides Sub MatchProperties(ByVal _from As VectorDraw.Professional.vdObjects.vdPrimary, ByVal thisdocument As VectorDraw.Professional.vdObjects.vdDocument)
            MyBase.MatchProperties(_from, thisdocument)
            Dim from As VectorDrawPolygon = CType(_from, VectorDrawPolygon)
            If from Is Nothing Then
                Return
            End If
            TextString = from.TextString
            NumSides = from.NumSides
            Radius = from.Radius
            TextStyle = from.TextStyle
            TextHeight = from.TextHeight
        End Sub
        Public Overrides Sub Transformby(ByVal mat As Matrix)
            If mat.IsUnitMatrix() Then
                Return
            End If
            Dim mult As Matrix = (ECSMatrix * mat)
            Dim matprops As MatrixProperties = mult.Properties
            Radius *= matprops.GetScaleXY()
            TextHeight *= matprops.GetScaleXY()
            MyBase.Transformby(mat)
        End Sub
        Public Overrides Function GetGripPoints() As gPoints
            Dim ret As gPoints = New gPoints()
            Dim cen As gPoint = New gPoint()
            ret.Add(cen)
            Dim stepangle As Double = Globals.VD_TWOPI / Me.NumSides
            Dim sang As Double = 0D
            Dim i As Integer
            For i = 0 To NumSides - 1 Step i + 1
                ret.Add(gPoint.Polar(cen, sang, Radius))
                sang += stepangle
            Next
            ECSMatrix.Transform(ret)
            Return ret
        End Function

        Public Overrides Sub MoveGripPointsAt(ByVal Indexes As Int32Array, ByVal dx As Double, ByVal dy As Double, ByVal dz As Double)
            If Indexes Is Nothing Or Indexes.Count = 0 Or (dx = 0.0 And dy = 0.0 And dz = 0.0) Then
                Return
            End If
            Dim mat As Matrix = New Matrix()
            mat.TranslateMatrix(dx, dy, dz)
            Dim grips As gPoints = GetGripPoints()
            ECSMatrix.GetInvertion().Transform(grips)
            If Indexes.Count = grips.Count Then

                Transformby(mat)
            Else
                Dim index As Integer
                For Each index In Indexes
                    Select Case index
                        Case 0
                            Transformby(mat)
                            Exit For
                        Case Else
                            Dim grip As gPoint = New gPoint(grips(index))
                            grip = ECSMatrix.Transform(grip)
                            grip += New gPoint(dx, dy, dz)
                            Me.Radius = gPoint.Distance2D(grip, Me.Origin)
                            Exit For
                    End Select
                Next
            End If
            Update()
    End Sub

    Public Overrides Function IsTableObjectDependOn(ByVal table As VectorDraw.Professional.vdObjects.vdPrimary) As Boolean
        If Deleted = True Then Return False ''always check if this object is deleted
        If MyBase.IsTableObjectDependOn(table) = True Then Return True ''always check the base implamentation
        ''check properties of object that reference the table object.
        If Not TextStyle Is Nothing And Object.ReferenceEquals(TextStyle, table) = True Then Return True
        Return False
    End Function
#End Region

End Class
Public Class ActionPolygon
    Inherits ActionEntityEx
    Private figure As VectorDrawPolygon
    Public Sub New(ByVal NumSides As Integer, ByVal text As String, ByVal textheight As Double, ByVal reference As gPoint, ByVal layout As vdLayout)
        MyBase.New(reference, layout)
        figure = New VectorDrawPolygon()
        figure.SetUnRegisterDocument(layout.Document)
        figure.setDocumentDefaults()
        figure.Origin = reference
        figure.Radius = 0D
        figure.TextHeight = textheight
        figure.TextString = text
        figure.NumSides = NumSides
    End Sub
    Protected Overrides Sub OnMyPositionChanged(ByVal NewPosition As gPoint)
        figure.Radius = NewPosition.Distance3D(figure.Origin)
    End Sub
    Public Overrides ReadOnly Property Entity() As vdFigure
        Get
            Return figure
        End Get
    End Property
    Public Shared Sub CmdPolygon(ByVal doc As vdDocument)
        Dim cen As gPoint = New gPoint()
        doc.Prompt("Origin Point : ")
        Dim ret As Object = doc.ActionUtility.getUserPoint()
        doc.Prompt(Nothing)
        If ret Is Nothing Or Not (TypeOf ret Is gPoint) Then
            Return
        End If
        cen = CType(ret, gPoint)
        doc.Prompt("Textstring : ")
        Dim text As String = doc.ActionUtility.getUserString()
        doc.Prompt(Nothing)
        If text Is Nothing Then
            Return
        End If
        doc.Prompt("Textheight : ")
        ret = doc.ActionUtility.getUserDist(cen)
        doc.Prompt(Nothing)
        If ret Is Nothing Then
            Return
        End If
        Dim textheight As Double = CType(ret, Double)
        Dim numsides As Integer = 0
        doc.Prompt("NumSides : ")
        Dim scode As StatusCode = doc.ActionUtility.getUserInt(numsides)
        doc.Prompt(Nothing)
        If scode <> VectorDraw.Actions.StatusCode.Success Then
            Return
        End If
        doc.Prompt("Radius : ")
        Dim aFig As ActionPolygon = New ActionPolygon(numsides, text, textheight, cen, doc.ActiveLayOut)
        doc.ActionAdd(aFig)
        scode = aFig.WaitToFinish()
        doc.Prompt(Nothing)
        If scode <> VectorDraw.Actions.StatusCode.Success Then
            Return
        End If
        aFig.Entity.Transformby(doc.User2WorldMatrix)
        doc.ActionLayout.Entities.AddItem(aFig.Entity)
        doc.ActionDrawFigure(aFig.Entity)
        Return
    End Sub
End Class

File: CustomObjects/vdPolygon/VectorDrawPolygon.cs
using System;
using System.Runtime.InteropServices;
using System.ComponentModel;
using VectorDraw.Serialize;
using VectorDraw.Professional;
using VectorDraw.Professional.vdPrimaries;
using VectorDraw.Professional.vdFigures;
using VectorDraw.Professional.vdCollections;
using VectorDraw.Professional.vdObjects;
using VectorDraw.Professional.Constants;
using VectorDraw.Geometry;
using VectorDraw.Render;
using VectorDraw.Professional.Actions;
using VectorDraw.Actions;

//This custom object also demonstrates how you can serialize an array of objects in vdml/vdcl format

//the VectorDrawCustomFigs.dll must be in the folder of vdrawcomponents
//or the supportpath must contain the path of the assemply dll
namespace VectorDrawCustomFigs
{
    public class VectorDrawPolygon : vdShape
    {
        #region main variables
        private vdTextstyle mTextStyle = null;
        private int mNumSides = 4;
        private string mText = "";
        private double mTextHeight = 1.0d;
        private double mRadius = 1.0d;
        private VectorDraw.Generics.vdArray<double> mMyArray = new VectorDraw.Generics.vdArray<double>();  
        #endregion

        #region ctor
        public VectorDrawPolygon()
        {
        }
        public VectorDrawPolygon(vdDocument doc)
        {
            SetUnRegisterDocument(doc);
            setDocumentDefaults();
        }
        #endregion
        #region not vdShape supported properties
        [Browsable(false)]
        public override double Rotation
        {//do not support vdShape Rotation
            get
            {
                //return base.Rotation;
                return 0.0d;
            }
            set
            {
                //base.Rotation = value;
            }
        }
        [Browsable(false)]
        public override Vector Scales
        {//do not support vdShape Scales
            get
            {
                //return base.Scales;
                return new Vector(1.0d, 1.0d, 1.0d);
            }
            set
            {
                //base.Scales = value;
            }
        }
        #endregion
        #region vdShape override properties
        public override gPoint Origin
        {
            get
            {
                return base.Origin;
            }
            set
            {
                base.Origin = value;
                Update();
            }
        }
        #endregion
        #region properties
        //You should use the AddHistory command in your properties so any change will be writen to the history
        //The string in the AddHistory command is the name of the property exactly and should be called before the property change.
        public int NumSides
        {
            get { return mNumSides; }
            set
            {
                AddHistory("NumSides", value);
                mNumSides = value;
            }
        }
        public string TextString
        {
            get { return mText; }
            set
            {
                AddHistory("TextString", value);
                mText = value;
            }
        }
        public double Radius
        {
            get { return mRadius; }
            set
            {
                AddHistory("Radius", value);
                mRadius = value;
            }
        }
        public vdTextstyle TextStyle
        {
            get
            {
                return mTextStyle;
            }
            set
            {
                AddHistory("TextStyle", value);
                mTextStyle = value;
            }
        }
        public double TextHeight
        {
            get
            {
                return mTextHeight;
            }
            set
            {
                AddHistory("TextHeight", value);
                mTextHeight = value;
            }
        }

        public void AddSomeValues()
        {
            mMyArray.AddItem(2.0);
            mMyArray.AddItem(3.0);
            mMyArray.AddItem(4.0);
            mMyArray.AddItem(5.0);
        }

        [TypeConverterAttribute()]
        [EditorAttribute(typeof(VectorDraw.Generics.vdCollectionDialogEditorNOSorted), typeof(System.Drawing.Design.UITypeEditor))]
        [Browsable(true)]
        public VectorDraw.Generics.vdArray<double> MyArray
        {
            get { return mMyArray; }
            set { mMyArray = value; }
        }
        #endregion
        #region serialization
        //Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
        //You can set value of custom property to any of the following type values: 
        //     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise

        public override void Serialize(Serializer serializer)
        {
            base.Serialize(serializer);
            serializer.Serialize(mNumSides, "NumSides");
            serializer.Serialize(mRadius, "Radius");
            serializer.Serialize(mText, "TextString");
            serializer.Serialize(mTextStyle, "TextStyle");
            serializer.Serialize(mTextHeight, "TextHeight");
            serializer.Serialize(MyArray.Count, "_Count");
            foreach (double obj in MyArray)
            {
                serializer.Serialize(obj, "_Item");
            }

        }
        public override bool DeSerialize(DeSerializer deserializer, string fieldname, object value)
        {
            if (base.DeSerialize(deserializer, fieldname, value)) return true;
            else if (fieldname == "NumSides") mNumSides = (int)value;
            else if (fieldname == "Radius") mRadius = (double)value;
            else if (fieldname == "TextString") mText = (string)value;
            else if (fieldname == "TextStyle") mTextStyle = (vdTextstyle)value;
            else if (fieldname == "TextHeight") mTextHeight = (double)value;
            else if (fieldname == "_Item" && value != null) MyArray.AddItem((double)value);
            else return false;
            return true;
        }
        #endregion
        #region curve calulation
        public override void FillShapeEntities(ref vdEntities entities)
        {//calculate shape entities in ecs 
            
                gPoint cen = new gPoint();
                vdText text = new vdText();
                entities.AddItem(text);
                text.MatchProperties(this, Document);

                text.Style = TextStyle;
                text.TextString = TextString;
                text.VerJustify = VdConstVerJust.VdTextVerCen;
                text.HorJustify = VdConstHorJust.VdTextHorCenter;
                text.Height = TextHeight;
                text.InsertionPoint = cen ;

                vdPolyline pl = new vdPolyline();
                entities.AddItem(pl);
                pl.MatchProperties(this, Document);
                Vertexes verts = new Vertexes();
                double stepangle = Globals.VD_TWOPI / this.NumSides;
                double sang = 0.0d;
                
                for (int i = 0; i < NumSides; i++)
                {
                    verts.Add(gPoint.Polar(cen, sang, Radius));
                    sang += stepangle;
                }
                pl.VertexList = verts;
                pl.Flag = VdConstPlineFlag.PlFlagCLOSE;
           
        }
        #endregion
        #region overrides
        protected override void OnDocumentDefaults()
        {
            base.OnDocumentDefaults();
            mTextStyle = Document.ActiveTextStyle;
            if (mTextStyle == null) mTextStyle = Document.TextStyles.Standard;
        }
        public override void InitializeProperties()
        {
            mNumSides = 4;
            mText = "";
            mTextHeight = 1.0d;
            mTextStyle = null;
            if (Document != null) mTextStyle = Document.TextStyles.Standard;
            mRadius = 1.0d;
            base.InitializeProperties();
        }
        public override void MatchProperties(VectorDraw.Professional.vdObjects.vdPrimary _from, VectorDraw.Professional.vdObjects.vdDocument thisdocument)
        {
            base.MatchProperties(_from, thisdocument);
            VectorDrawPolygon from = _from as VectorDrawPolygon;
            if (from == null) return;
            TextString = from.TextString;
            NumSides = from.NumSides;
            Radius = from.Radius;
            TextStyle = from.TextStyle;
            TextHeight = from.TextHeight;
        }
        public override void Transformby(Matrix mat)
        {
            if (mat.IsUnitMatrix()) return;
            Matrix mult = (ECSMatrix * mat);
            MatrixProperties matprops = mult.Properties;
            Radius *= matprops.GetScaleXY();
            TextHeight *= matprops.GetScaleXY();
            base.Transformby(mat);
        }
        public override gPoints GetGripPoints()
        {
            gPoints ret = new gPoints();
            gPoint cen = new gPoint();
            ret.Add(cen);
            double stepangle = Globals.VD_TWOPI / this.NumSides;
            double sang = 0.0d;
            for (int i = 0; i < NumSides; i++)
            {
                ret.Add(gPoint.Polar(cen, sang, Radius));
                sang += stepangle;
            }
            ECSMatrix.Transform(ret);
            return ret;
        }
        public override void MoveGripPointsAt(Int32Array Indexes, double dx, double dy, double dz)
        {
            if (Indexes == null || Indexes.Count == 0 || (dx == 0.0 && dy == 0.0 && dz == 0.0)) return;
            Matrix mat = new Matrix();
            mat.TranslateMatrix(dx, dy, dz);
            gPoints grips = GetGripPoints();
            ECSMatrix.GetInvertion().Transform(grips);
            if (Indexes.Count == grips.Count)
            {

                Transformby(mat);
            }
            else
            {
                foreach (int index in Indexes)
                {
                    switch (index)
                    {
                        case 0:
                            Transformby(mat);
                            break;
                        default:
                            gPoint grip = new gPoint(grips[index]);
                            grip = ECSMatrix.Transform(grip);
                            grip += new gPoint(dx, dy, dz);
                            this.Radius = gPoint.Distance2D(grip, this.Origin);
                            break;
                    }
                }
            }
            Update();
        }
        public override bool IsTableObjectDependOn(vdPrimary table)
        {
            if (Deleted) return false;//always check if this object is deleted
            if (base.IsTableObjectDependOn(table)) return true;//always check the base implamentation
            //check properties of object that reference the table object.
            if (TextStyle != null && object.ReferenceEquals(TextStyle,table)) return true;
            return false;
        }
        public override void GetTableDependecies(vdTableDependeciesArgs args)
        {
            base.GetTableDependecies(args);
            if (args.IsObjectFound) return;
            if (Deleted) return;
            args.AddItem(TextStyle);
            if (TextStyle != null) TextStyle.GetTableDependecies(args);
        }
        #endregion
    }
    public class ActionPolygon : ActionEntity
    {
        private VectorDrawPolygon figure;
        public ActionPolygon(int NumSides, string text, double textheight, gPoint reference, vdLayout layout)
            : base(reference, layout)
        {
            ValueTypeProp |= valueType.DISTANCE;
            figure = new VectorDrawPolygon();
            figure.SetUnRegisterDocument(layout.Document);
            figure.setDocumentDefaults();
            figure.Origin = reference;
            figure.Radius = 0.0d;
            figure.TextHeight = textheight;
            figure.TextString = text;
            figure.NumSides = NumSides;
        }
        protected override void OnMyPositionChanged(gPoint newPosition)
        {
            figure.Radius = newPosition.Distance3D(figure.Origin);
        }
        public override vdFigure Entity
        {
            get
            {
                return figure;
            }
        }
        public static void CmdPolygon(vdDocument doc)
        {
            gPoint cen = new gPoint();
            doc.Prompt("Origin Point : ");
            object ret = doc.ActionUtility.getUserPoint();
            doc.Prompt(null);
            if (ret == null || !(ret is gPoint)) goto error;
            cen = ret as gPoint;
            doc.Prompt("Textstring : ");
            string text = doc.ActionUtility.getUserString();
            doc.Prompt(null);
            if (text == null) goto error;
            doc.Prompt("Textheight : ");
            ret = doc.ActionUtility.getUserDist(cen);
            doc.Prompt(null);
            if (ret == null) goto error;
            double textheight = (double)ret;
            int numsides = 0;
            doc.Prompt("NumSides : ");
            StatusCode scode = doc.ActionUtility.getUserInt(out numsides);
            doc.Prompt(null);
            if (scode != VectorDraw.Actions.StatusCode.Success) goto error;
            doc.Prompt("Radius : ");
            ActionPolygon aFig = new ActionPolygon(numsides, text, textheight, cen, doc.ActiveLayOut);
            doc.ActionAdd(aFig);
            scode = aFig.WaitToFinish();
            doc.Prompt(null);
            if (scode != VectorDraw.Actions.StatusCode.Success) goto error;
            aFig.Entity.Transformby(doc.User2WorldMatrix);
            doc.ActionLayout.Entities.AddItem(aFig.Entity);
            doc.ActionDrawFigure(aFig.Entity);
            return;
        error:
            return;
        }
    }
}
